Add new VideoRangeTypes to fully support DoVi on webOS

viktory36 opened the pull request:
This commit adds two new values to the VideoRangeType enum to talk about media that have both DOVI and HDR10 or HLG data (DoVi Profile 8) to enable Dolby Vision playback on those browsers.

**Changes**
Introduced `DOVI_with_HDR10_Fallback` and `DOVI_with_HLG_Fallback` VideoRangeType enums, and added references to these new types wherever I found `VideoRangeType.HDR10 / HLG`.
MediaStream.cs will now test both of it's HDR10/HLG and DOVI conditions before returning the video range type.

**Issues**
Addresses #10468

viktory36 commented:
The build pipeline does not like the name `DOVI_with_HDR10_Fallback` as they have underscores. Do you have other suggestions for the enum name?

`
/home/vsts/work/1/s/Jellyfin.Data/Enums/VideoRangeType.cs(36,5): error CA1707: Remove the underscores from member name Jellyfin.Data.Enums.VideoRangeType.DOVI_with_HDR10_Fallback (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1707) [/home/vsts/work/1/s/Jellyfin.Data/Jellyfin.Data.csproj]
/home/vsts/work/1/s/Jellyfin.Data/Enums/VideoRangeType.cs(41,5): error CA1707: Remove the underscores from member name Jellyfin.Data.Enums.VideoRangeType.DOVI_with_HLG_Fallback (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1707) [/home/vsts/work/1/s/Jellyfin.Data/Jellyfin.Data.csproj]
`

crobibero commented:
> The build pipeline does not like the name DOVI_with_HDR10_Fallback as they have underscores. Do you have other suggestions for the enum name?


Maybe just DOVIHDR10 and DOVIHLG?

it's more difficult since both are acronyms..

viktory36 commented:
> > The build pipeline does not like the name DOVI_with_HDR10_Fallback as they have underscores. Do you have other suggestions for the enum name?
> 
> Maybe just `DOVIHDR10` and `DOVIHLG`?
> 
> it's more difficult since both are acronyms..

Yup. I was thinking more of what the end user might see - if they navigate over to the playback info screen and find "DOVIHDR10" or "DOVIProfile8" I think that would be more confusing than the current.

DOVIwithHDR10
DOVIwithHLG

DoViHDR10
DoViHLG

I think I like this last one DoViHDR10 better, but not very much - don't want to mislead any users into thinking they're seeing DOVI when it's HDR.

Would you be opposed to disabling this warning CA1707? Would throwing that into the jellyfin.ruleset work for the build pipeline to pass?

viktory36 commented:
Changed it to `DolbyVisionWithHDR10Fallback`. The only failing check now also failed for the initial PR introducing VideoRangeType enums. Not sure what needs to change to correct that.

I also reverted the "[Dont assume HLG/HDR10 tonemapper can support DOVI+HLG/HDR10](https://github.com/jellyfin/jellyfin/pull/10469/commits/d330436719ba2b567bcb948b86de1f329351c650)" commit - it makes sense that if the DOVI+HDR10 media were previously reported as HDR10 and had no issues with any HwTonemap, the new range types representing this media should fall in the same bucket.

viktory36 commented:
Dear @nyanmisaka , please let me know your thoughts on this?

nyanmisaka commented on Jellyfin.Data/Enums/VideoRangeType.cs:
```diff
@@ -30,6 +30,16 @@ public enum VideoRangeType
+    /// <summary>
+    /// Dolby Vision with HLG video range fallback (12bit or 10bit).
+    /// </summary>
+    DolbyVisionWithHLGFallback,
```
We can shorten its name a little bit.

```cs
    /// <summary>
    /// Dolby Vision video range type without any fallback (12bit).
    /// </summary>
    DOVI,

    /// <summary>
    /// Dolby Vision with HDR10 video range type fallback (12bit or 10bit).
    /// </summary>
    DOVIWithHDR10,

    /// <summary>
    /// Dolby Vision with HLG video range type fallback (12bit or 10bit).
    /// </summary>
    DOVIWithHLG,
```

nyanmisaka commented on MediaBrowser.Model/Entities/MediaStream.cs:
```diff
@@ -712,9 +704,45 @@ public bool SupportsSubtitleConversionTo(string toCodec)
+                else if (isHLG)
+                {
+                    return (VideoRange.HDR, VideoRangeType.DolbyVisionWithHLGFallback);
+                }
```
IMO this is not that reliable. We have `dvBlCompatId` that can be used to identify the DOVI fallback types.

https://github.com/jellyfin/jellyfin/blob/8f2829e1aa696d37d338fceed2ea1d9edf1ca824/MediaBrowser.Model/Entities/MediaStream.cs#L206-L208

nyanmisaka commented:
> @nyanmisaka I have some changes locally that implement the suggestions above, but I can't push to the pull request branch. Should I open a new pull request, as the author seems inactive? @viktory36 I think there is an option to permit that, if you would be so kind as to enable it.

If the OP still doesn't respond, you can make your branch based on his commits and open a new PR.

viktory36 commented:
Hey @GeorgeH005, I've enabled "allow edits by maintainers" to this PR, and added you as a collaborator to my fork if direct commits to the pr doesn't work.
If that's still not enough to be able to add a commit to this PR, please feel free to raise another.

Thanks and excited to have this on main!

nyanmisaka commented on Jellyfin.Api/Controllers/DynamicHlsController.cs:
```diff
@@ -1804,6 +1804,8 @@ private string GetVideoArguments(StreamState state, int startNumber, bool isEven
             if (EncodingHelper.IsCopyCodec(codec)
                 && (state.VideoStream.VideoRangeType == VideoRangeType.DOVI
+                    || state.VideoStream.VideoRangeType == VideoRangeType.DOVIWithHDR10
+                    || state.VideoStream.VideoRangeType == VideoRangeType.DOVIWithHLG
```
Some clients do not support `dvh1` FourCC but they support HDR10 & HLG in `hvc1`, such as the HLS.js used in non-Safari browsers. Please refer to the [code here](https://github.com/jellyfin/jellyfin/blob/fab3151679999bddf767f8c1e5e0826f58f63e4f/MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs#L2200-L2212) to set the appropriate video tag according to the VideoRangeType reported by the client.

nyanmisaka commented on MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:
```diff
@@ -6118,7 +6121,9 @@ public string GetVideotoolboxVidDecoder(EncodingJobInfo state, EncodingOptions o
                                     && (state.VideoStream.VideoRangeType == VideoRangeType.HDR10
                                         || state.VideoStream.VideoRangeType == VideoRangeType.HLG
-                                        || (state.VideoStream.VideoRangeType == VideoRangeType.DOVI
+                                        || ((state.VideoStream.VideoRangeType == VideoRangeType.DOVI
+                                            || state.VideoStream.VideoRangeType == VideoRangeType.DOVIWithHDR10
+                                            || state.VideoStream.VideoRangeType == VideoRangeType.DOVIWithHLG)
```
```suggestion
                                        || (state.VideoStream.VideoRangeType == VideoRangeType.DOVI
```

In VideoToolbox the `useOclToneMapping` is only enabled for DOVI Profile 5. No need to change here.

gnattu replied:
I *think* it could be needed, as the user might have explicitly disabled VideoToolbox's native tone mapping. If `VideoRangeType.DOVI` does not imply `VideoRangeType.DOVIWithHDR10` or `VideoRangeType.DOVIWithHLG`, then users with such configurations might not have hwsurface initialized, which could cause hwmap to fail.

gnattu replied:
> If so there is no reason to exclude VideoRangeType.{HDR10,HLG}.
They are already included, the code segment shown here might be incomplete, but all kind of HDR is already included
https://github.com/jellyfin/jellyfin/blob/6b4951fcf5c0fe2a473b2c13367fdf367a4a4c78/MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs#L6121-L6128

nyanmisaka replied:
Then this is necessary. I read the wrong line.

nyanmisaka reviewed (changes requested):
When everything is resolved, this should be merged after the relevant changes in jellyfin-web are merged. Merging this one alone will cause regression.

viktory36 commented:
@GeorgeH005 , there's a PR open for web here with an open review comment: https://github.com/jellyfin/jellyfin-web/pull/4916

Let me take care of that one

nyanmisaka commented on Jellyfin.Api/Controllers/DynamicHlsController.cs:
```diff
@@ -1797,13 +1797,17 @@ private string GetVideoArguments(StreamState state, int startNumber, bool isEven
                 && (state.VideoStream.VideoRangeType == VideoRangeType.DOVI
+                    || (state.VideoStream.VideoRangeType == VideoRangeType.DOVIWithHDR10 && requestedRange.Contains(VideoRangeType.DOVIWithHDR10.ToString(), StringComparison.OrdinalIgnoreCase))
+                    || (state.VideoStream.VideoRangeType == VideoRangeType.DOVIWithHLG && requestedRange.Contains(VideoRangeType.DOVIWithHLG.ToString(), StringComparison.OrdinalIgnoreCase))
                     || string.Equals(state.VideoStream.CodecTag, "dovi", StringComparison.OrdinalIgnoreCase)
                     || string.Equals(state.VideoStream.CodecTag, "dvh1", StringComparison.OrdinalIgnoreCase)
                     || string.Equals(state.VideoStream.CodecTag, "dvhe", StringComparison.OrdinalIgnoreCase)))
```
We can drop these now.
```suggestion
```

nyanmisaka replied:
ffmpeg currently does not support those profiles (handling metdata)... Profile 10 for AV1 and Profile 10 for 3D.

nyanmisaka replied:
Currently, the support for DV in FFmpeg is limited to P5/7/8 implemented through reverse engineering. Additional metadata will be ignored, so setting dvh1 here won't help.

nyanmisaka commented:
> @nyanmisaka According to our conversation over in [jellyfin/jellyfin-web#5282](https://github.com/jellyfin/jellyfin-web/pull/5282), should I adjust the code to only tag profiles 5 and 8 as dolby vision, and the rest as regular hdr/hlg or sdr?

DV P5/8 => (HDR, DOVI*)
DV P7 => (HDR, HDR10) // no HLG in P7, i think?
Other DV => (SDR, SDR)

nyanmisaka commented:
> Should I also check if the client has requested DOVI as a VideoRangeType before tagging the stream as dvh1? Profile 5 will look wrong in both cases if dvh1 isn't supported ( I think ), but Profile 8.4 if tagged as hvc1 should fallback to SDR, which seems the better behaviour to me. These are the only two cases that I think this applies to

Just tested and found we should move that `requestedRange.Contains` to `Encodinghelper->CanStreamCopyVideo()`
https://github.com/jellyfin/jellyfin/blob/b6d130ae2d337294904f6d46d943e9f82c162978/MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs#L2200-L2212

And enable `dvh1` for all DOVI*. Otherwise this will prevent remuxing and trigger transcoding.

nyanmisaka commented:
> @nyanmisaka That makes sense, I assume that if CanStreamCopyVideo is false then OutputVideoCodec is false, so EncodingHelper.IsCopyCodec is also false, so those checks in GetVideoArguments are redundant anyways, correct?

Yes. When the program runs to GetVideoArguments ->(IsCopyCodec && is dvh1), this means that the client and the server negotiate and prepare for remuxing instead of transcoding.

Just allow the server to enable stream copy of DV P8 for clients (Chrome/hls.js) that only support HDR10/HLG in `CanStreamCopyVideo()`.

nyanmisaka commented:
> Just allow the server to enable stream copy of DV P8 for clients (Chrome/hls.js) that only support HDR10/HLG in `CanStreamCopyVideo()`.

No. ^^^

nyanmisaka commented:
> I thought previously that tagging Profile 8.4 (DOVI + SDR) as DOVI would be correct, but since the stream isn't going to be copied in case DOVI is not supported, there is going to be unnecessary transcoding/tonemapping. Should I introduce a new type for 8.4? Handling it as SDR in the first place doesn't seem like a good idea.

```cs
                return dvProfile switch
                {
                    5 => (VideoRange.HDR, VideoRangeType.DOVI),
                    8 => dvBlCompatId switch
                    {
                        1 => (VideoRange.HDR, VideoRangeType.DOVIWithHDR10),
                        4 => (VideoRange.HDR, VideoRangeType.DOVIWithHLG), //< you mean this ? profile 8.4 HLG
                        _ => (VideoRange.HDR, VideoRangeType.DOVI)
                    },
                    7 => (VideoRange.HDR, VideoRangeType.HDR10),
                    _ => (VideoRange.SDR, VideoRangeType.SDR)
                };
```

Which profile is SDR?

nyanmisaka commented:
> > And enable `dvh1` for all DOVI*. Otherwise this will prevent remuxing and trigger transcoding.
> > If Chrome does not report DOVIWithHDR10 support and the video file is DV P8, then CanStreamCopyVideo() will return false.
> 
> @nyanmisaka But then the stream is going to be tagged as dvh1 even with Chrome as a client, and based on previous messages my understanding is that for the fallback to work, the stream needs to be hvc1.
> 
> > Just tested and found we should move that requestedRange.Contains to Encodinghelper->CanStreamCopyVideo()
> 
> So I think that check needs to exist in both places, not be moved

Yeah. You are right. I take it too easy.

nyanmisaka commented on MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:
```diff
@@ -2205,7 +2208,10 @@ public bool CanStreamCopyVideo(EncodingJobInfo state, MediaStream videoStream)
-                if (!requestedRangeTypes.Contains(videoStream.VideoRangeType.ToString(), StringComparison.OrdinalIgnoreCase))
+                // DOVIWithHDR10 should be compatible with HDR10 supporting players. Same goes with HLG. So allow copy of those formats
+                if (!requestedRangeTypes.Contains(videoStream.VideoRangeType.ToString(), StringComparison.OrdinalIgnoreCase)
+                     && !((requestedRangeTypes.Contains("HDR10", StringComparison.OrdinalIgnoreCase) && string.Equals(videoStream.VideoRangeType.ToString(), "DOVIWithHDR10", StringComparison.OrdinalIgnoreCase))
+                            || (requestedRangeTypes.Contains("HLG", StringComparison.OrdinalIgnoreCase) && string.Equals(videoStream.VideoRangeType.ToString(), "DOVIWithHLG", StringComparison.OrdinalIgnoreCase))))
```
```suggestion
                     && !((requestedRangeTypes.Contains("HDR10", StringComparison.OrdinalIgnoreCase) && videoStream.VideoRangeType == VideoRangeType.DOVIWithHDR10))
                            || (requestedRangeTypes.Contains("HLG", StringComparison.OrdinalIgnoreCase) && videoStream.VideoRangeType == VideoRangeType.DOVIWithHLG)))
```

nyanmisaka commented:
> @nyanmisaka My mistake, Profile 8.2. Sorry about that

This should be okay? We only care about HDR here. So just filter out the `dvBlCompatId` that are known to contain HDR compatible layers.
```
                    8 => dvBlCompatId switch
                    {
                        1 => (VideoRange.HDR, VideoRangeType.DOVIWithHDR10),
                        4 => (VideoRange.HDR, VideoRangeType.DOVIWithHLG), //< you mean this ? profile 8.4 HLG
                        _ => (VideoRange.SDR, VideoRangeType.SDR)
                    },
```

nyanmisaka commented:
Then I think you can add a `DOVIWithSDR`. Even though I don't have such a clip I know it exists.

nyanmisaka commented:
So it's an enhanced SDR video, so it *should* be (SDR, DOVIWithSDR).
It only makes sense during stream-copy. In other cases, there is no difference from traditional SDR.

nyanmisaka commented:
> @nyanmisaka Did some more research and found out that Apple's HLS spec supports profiles 5 and 8, but mpeg-ts should support dual layer 7 and every other available profile, at least according to Dolby. Not sure if we want to change our implementation now, at least when it comes to Profile 7, as I cannot speak at all for client support. I know we already had a conversation about Profile 7, but this is news to me. Given the option "Prefer fmp4-hls container..." in the playback settings, I assumed that there were mpeg-ts streaming capabilities. If I am wrong, of course disregard the above comment.

Apple just rejects the HEVC codec in MPEG-TS HLS, so this won't happen.

nyanmisaka commented on Jellyfin.Data/Enums/VideoRangeType.cs:
```diff
@@ -26,10 +26,25 @@ public enum VideoRangeType
 
     /// <summary>
-    /// Dolby Vision video range type (12bit).
+    /// Dolby Vision video range type (10bit).
```
Profile 5 uses ICtCp instead of YCbCr, which is more efficient. So it's 10-Bit encoded/12-bit remapped.

nyanmisaka commented on MediaBrowser.Model/Entities/MediaStream.cs:
```diff
@@ -707,34 +707,46 @@ public bool SupportsSubtitleConversionTo(string toCodec)
 
             var isDoViHDRProfile = dvProfile == 5 || dvProfile == 7 || dvProfile == 8;
-            var isDoViHDRFlag = rpuPresentFlag && blPresentFlag && (dvBlCompatId == 0 || dvBlCompatId == 1 || dvBlCompatId == 4);
+            var isDoViHDRFlag = rpuPresentFlag && blPresentFlag && (dvBlCompatId == 0 || dvBlCompatId == 1 || dvBlCompatId == 4 || dvBlCompatId == 2 || dvBlCompatId == 6);
```
> dvBlCompatId == 2 || dvBlCompatId == 6

It's actually `isDoViSDRFlag`.

So just rename:
`isDoViHDRProfile` to `isDoViProfile`
`isDoViHDRFlag` to `isDoViFlag`

nyanmisaka commented on Jellyfin.Api/Controllers/DynamicHlsController.cs:
```diff
@@ -1797,16 +1797,18 @@ private string GetVideoArguments(StreamState state, int startNumber, bool isEven
 
         var args = "-codec:v:0 " + codec;
 
+        var requestedRange = state.GetRequestedRangeTypes(codec);
+
```
```suggestion
```

nyanmisaka commented on Jellyfin.Api/Controllers/DynamicHlsController.cs:
```diff
@@ -1797,16 +1797,18 @@ private string GetVideoArguments(StreamState state, int startNumber, bool isEven
             || string.Equals(state.ActualOutputVideoCodec, "hevc", StringComparison.OrdinalIgnoreCase)
             || string.Equals(codec, "h265", StringComparison.OrdinalIgnoreCase)
             || string.Equals(codec, "hevc", StringComparison.OrdinalIgnoreCase))
         {
```
```suggestion
        {
            var types = state.GetRequestedRangeTypes(codec);
            var haveDOVI = types.Contains(VideoRangeType.DOVI.ToString(), StringComparison.OrdinalIgnoreCase);
            var haveDOVIWithHDR10 = types.Contains(VideoRangeType.DOVIWithHDR10.ToString(), StringComparison.OrdinalIgnoreCase);
            var haveDOVIWithHLG = types.Contains(VideoRangeType.DOVIWithHLG.ToString(), StringComparison.OrdinalIgnoreCase);
            var haveDOVIWithSDR = types.Contains(VideoRangeType.DOVIWithSDR.ToString(), StringComparison.OrdinalIgnoreCase);

```

nyanmisaka commented on MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:
```diff
@@ -2205,7 +2208,11 @@ public bool CanStreamCopyVideo(EncodingJobInfo state, MediaStream videoStream)
                 }
 
-                if (!requestedRangeTypes.Contains(videoStream.VideoRangeType.ToString(), StringComparison.OrdinalIgnoreCase))
+                // DOVIWithHDR10 should be compatible with HDR10 supporting players. Same goes with HLG and of course SDR. So allow copy of those formats
+                if (!requestedRangeTypes.Contains(videoStream.VideoRangeType.ToString(), StringComparison.OrdinalIgnoreCase)
+                     && !((requestedRangeTypes.Contains("HDR10", StringComparison.OrdinalIgnoreCase) && videoStream.VideoRangeType == VideoRangeType.DOVIWithHDR10)
+                            || (requestedRangeTypes.Contains("HLG", StringComparison.OrdinalIgnoreCase) && videoStream.VideoRangeType == VideoRangeType.DOVIWithHLG)
+                            || (requestedRangeTypes.Contains("SDR", StringComparison.OrdinalIgnoreCase) && videoStream.VideoRangeType == VideoRangeType.DOVIWithSDR)))
```
```suggestion
                var haveDOVIWithHDR10 = requestedRangeTypes.Contains(VideoRangeType.DOVIWithHDR10.ToString(), StringComparison.OrdinalIgnoreCase);
                var haveDOVIWithHLG = requestedRangeTypes.Contains(VideoRangeType.DOVIWithHLG.ToString(), StringComparison.OrdinalIgnoreCase);
                var haveDOVIWithSDR = requestedRangeTypes.Contains(VideoRangeType.DOVIWithSDR.ToString(), StringComparison.OrdinalIgnoreCase);

                // DOVIWithHDR10 should be compatible with HDR10 supporting players. Same goes with HLG and of course SDR. So allow copy of those formats
                if (!requestedRangeTypes.Contains(videoStream.VideoRangeType.ToString(), StringComparison.OrdinalIgnoreCase)
                    && !((videoStream.VideoRangeType == VideoRangeType.DOVIWithHDR10 && haveDOVIWithHDR10)
                          || (videoStream.VideoRangeType == VideoRangeType.DOVIWithHLG && haveDOVIWithHLG)
                          || (videoStream.VideoRangeType == VideoRangeType.DOVIWithSDR && haveDOVIWithSDR)))
```

Each line is already too long, Extract the contains* to vars to increase readability.

nyanmisaka commented on Jellyfin.Api/Controllers/DynamicHlsController.cs:
```diff
@@ -1797,16 +1797,18 @@ private string GetVideoArguments(StreamState state, int startNumber, bool isEven
-                    || string.Equals(state.VideoStream.CodecTag, "dovi", StringComparison.OrdinalIgnoreCase)
-                    || string.Equals(state.VideoStream.CodecTag, "dvh1", StringComparison.OrdinalIgnoreCase)
-                    || string.Equals(state.VideoStream.CodecTag, "dvhe", StringComparison.OrdinalIgnoreCase)))
+                && ((state.VideoStream.VideoRangeType == VideoRangeType.DOVI && requestedRange.Contains(VideoRangeType.DOVI.ToString(), StringComparison.OrdinalIgnoreCase))
+                    || (state.VideoStream.VideoRangeType == VideoRangeType.DOVIWithHDR10 && requestedRange.Contains(VideoRangeType.DOVIWithHDR10.ToString(), StringComparison.OrdinalIgnoreCase))
+                    || (state.VideoStream.VideoRangeType == VideoRangeType.DOVIWithHLG && requestedRange.Contains(VideoRangeType.DOVIWithHLG.ToString(), StringComparison.OrdinalIgnoreCase))
+                    || (state.VideoStream.VideoRangeType == VideoRangeType.DOVIWithSDR && requestedRange.Contains(VideoRangeType.DOVIWithSDR.ToString(), StringComparison.OrdinalIgnoreCase))))
```
```suggestion
                && ((state.VideoStream.VideoRangeType == VideoRangeType.DOVI && haveDOVI)
                     || (state.VideoStream.VideoRangeType == VideoRangeType.DOVIWithHDR10 && haveDOVIWithHDR10)
                     || (state.VideoStream.VideoRangeType == VideoRangeType.DOVIWithHLG && haveDOVIWithHLG)
                     || (state.VideoStream.VideoRangeType == VideoRangeType.DOVIWithSDR && haveDOVIWithSDR)))
```

Each line is already too long, Extract the contains* to vars to increase readability.

nyanmisaka commented:
Just want to double check, have you also tested these PRs on real webOS hardware and Chrome with fMP4-HLS enabled?

nyanmisaka commented:
> The hev1 mp4 might be something to look into. When converting that I did not specify a "-tag:v:0 hev1" argument I just did a "-i 12345.mkv -c:v copy -c:a copy -sn 12345.mp4".

Are the source files mentioned here DV P8? What results did you expect?

viktory36 commented:
@GeorgeH005 nitpicked on the DOVIWithSDR enum description a little bit =] I reasoned SDR can commonly appear in 8bit or 10bit color depth, and if DV will be at 10bit depth, then (8bit / 10bit) is a fair statement to make -- though I haven't found a file that has DV with 8bit SDR fallback happening on the same media.

nyanmisaka commented:
> The source file was an mkv with Dolby Vision Profile 8 but I did not specify a tag when converting the file, so it might have been my fault

ffmpeg only passes DV metadata from demuxer to muxer when `-strict -2` is used.

nyanmisaka reviewed (approved):
Then I think this is good to go. The web side fixes should be merged first.

Considering this actually fixes Dolby Vision playback on webOS, IMO this isn't a new feature.

nyanmisaka commented:
> @nyanmisaka The jellyfin-web team believes this should be merged before theirs.

~~1. https://github.com/jellyfin/jellyfin-web/pull/4916~~
2. https://github.com/jellyfin/jellyfin/pull/10469
3. https://github.com/jellyfin/jellyfin-web/pull/5282

All right. Let's merge this first. cc @crobibero
